romtools/linalg/linalg.py

```diff
@@ -192,6 +192,142 @@ def _basic_max_via_python(a: np.ndarray, axis=None, comm=None):
+               51.
+               -24.
+               45.
+      =======================
+       rank 2  -4.
+
+    Suppose that we do:
+
+        res = la.argmax(a, comm)
+
+    then ALL ranks will contain res = (51., 1, 1).
+    (The global maximum is 51., and it occurs at index 1 of the local array on Rank 1.)
+
+    Example 2:
+    **********
+
+       rank 0  2.2  1.3  4.
+               3.3  5.0  33.
+      =======================
+       rank 1  40.  -2.  -4.
+               51.   4.   6.
+               -24.  8.   9.
+               45.  -3.  -4.
+      =======================
+       rank 2  -4.  8.   9.
+
+    Suppose that we do:
+
+       res = la.argmax(a, comm)
+
+    then ALL ranks will contain res = (51,. 3, 1)
+    (The maximum value is 51., and it occurs at index 3 of the flattened local array on Rank 1.)
+
+    Example 3:
+    **********
+
+       / 3.   4.   /  2.   8.   2.   1.   / 2.
+      /  6.  -1.  /  -2.  -1.   0.  -6.  /  0.    -> slice T(:,:,1)
+     /  -7.   5. /    5.   0.   3.   1. /   3.
+    |-----------|----------------------|--------
+    | 2.   3.   |  4.   5.  -2.   4.   | -4.
+    | 1.   5.   | -2.   4.   8.  -3.   |  8.    ->  slice T(:,:,0)
+    | 4.   3.   | -4.   6.   9.  -4.   |  9.
+
+        r0                r1              r2
+
+    Suppose that we do:
+
+        res = la.argmax(a, comm)
+
+    then ALL ranks will contain res = (9.0, 20, 1)
+    (The maximum value is 9.0. It occurs on both Rank 1 and Rank 2, but we automatically return the
+    index on the lowest rank. In this case, that is index 20 of the flattened local array on Rank 1.)
+
+    '''
+    # Enforce preconditions
+    assert a.ndim <= 3, "a must be at most a rank-3 tensor"
+
+    # Return "local" result if not running distributed
+    if comm is None or comm.Get_size() == 1:
+        return np.argmax(a)
+
+    # Get local array argmax result
+    a = a.flatten()
+    local_max_index = np.argmax(a)
+    local_max_val = a[local_max_index]
+
+    # Set up local solution
+    tmp = np.zeros(3)
+    tmp[0] = local_max_val
+    tmp[1] = local_max_index
+    tmp[2] = comm.Get_rank() if comm is not None else 0
+
+    # Otherwise, find distributed max index
+    import mpi4py
+    from mpi4py import MPI
+
+    # Define custom MPI op
+    def mycomp(A_mem,B_mem,dt):
+        # Get matrices from memory buffers
+        A = np.frombuffer(A_mem)
+        B = np.frombuffer(B_mem)
+
+        # Return the index of the max (or the max on the lowest rank, if multiple occurrences)
+        if A[0] < B[0] or (A[0] == B[0] and A[2] > B[2]):
+            result = B
+        else:
+            result = A
+
+        # Copy result to B for next comparison
+        B[:] = result
+
+    # Perform operation
+    result = np.zeros(3)
+    myop = MPI.Op.Create(mycomp, commute=False)
+    comm.Allreduce(tmp, result, op=myop)
+    myop.Free()
+
+    # Return val (float), index (int64), and rank (int)
+    return result[0], np.int64(result[1]), int(result[2])
```

Comment:
can you please change this AND above to only return index and rank, we don't really need the value so this should simplify above since the tmp array can be only int

Reply:
I changed the function to only return the index and rank, but we still need to hold onto the value in the `tmp` arrays so that we can perform the comparison in the allreduce.